File: Tutorials/Codes/System_Dynamics_Course/Chapter13/Lesson3/Chapter13_Lesson3.m
% Chapter13_Lesson3.m
% System Dynamics (Control Engineering) — Chapter 13, Lesson 3
% Modal Coordinates and Decoupling of MDOF Systems (Undamped Case)
%
% This script:
% 1) Builds a 3-DOF mass-spring chain (M, K)
% 2) Solves generalized EVP: K*phi = lambda*M*phi
% 3) Mass-normalizes: Phi'*M*Phi = I, Phi'*K*Phi = Omega^2
% 4) Simulates forced response in modal coordinates (decoupled ODEs)
% 5) Builds a simple Simulink model programmatically (optional)

clear; clc;

%% 1) Model
m1 = 1.0; m2 = 1.2; m3 = 0.9;
k1 = 1200; k2 = 900; k3 = 700; k4 = 1100;

M = diag([m1 m2 m3]);
K = [k1+k2, -k2,   0;
     -k2,  k2+k3, -k3;
      0,    -k3,  k3+k4];

%% 2) Modal decomposition (symmetric generalized eigenproblem)
[Phi,Lam] = eig(K,M);               % columns: eigenvectors
lam = diag(Lam);
[lam, idx] = sort(lam,'ascend');
Phi = Phi(:,idx);

omega = sqrt(max(lam,0));

% Mass-normalize columns: Phi'*M*Phi = I
for i=1:size(Phi,2)
    mi = Phi(:,i)'*M*Phi(:,i);
    Phi(:,i) = Phi(:,i)/sqrt(mi);
end

Omega2 = Phi'*K*Phi;
disp('Natural frequencies (rad/s):'); disp(omega');
disp('Check Phi''*M*Phi:'); disp(Phi'*M*Phi);
disp('Check Phi''*K*Phi:'); disp(Omega2);

%% 3) Forcing and decoupled simulation
F0 = 10;
Omega = 0.9*omega(1);

f = @(t)[F0*sin(Omega*t); 0; 0];      % physical force
p = @(t)Phi'*f(t);                     % modal force (mass-normalized)

% Modal ODE: qdd + diag(omega.^2) q = p(t)
n = 3;
ode = @(t,z)[ z(n+1:end);
             - (omega.^2).*z(1:n) + p(t) ];

tspan = [0 8];
z0 = zeros(2*n,1); % [q0; qd0]

opts = odeset('RelTol',1e-8,'AbsTol',1e-10);
[t,z] = ode45(ode,tspan,z0,opts);
q = z(:,1:n).';
x = Phi*q; % x(t) = Phi q(t)

figure; plot(t,x(1,:), 'LineWidth',1.2); grid on;
xlabel('t [s]'); ylabel('x_1 [m]'); title('Displacement of DOF 1 via modal simulation');

figure; plot(t,q.', 'LineWidth',1.2); grid on;
xlabel('t [s]'); ylabel('q_i'); legend('q_1','q_2','q_3');
title('Modal coordinates');

%% 4) Optional: build a Simulink model programmatically
% This creates a state-space block for the *decoupled* modal system:
%   [q; qd]' = A*[q; qd] + B*u,  with u(t) = sin(Omega t)
%   y = x1 = e1^T * Phi * q

make_simulink = true;
if make_simulink
    mdl = 'Chapter13_Lesson3_Simulink';
    if bdIsLoaded(mdl); close_system(mdl,0); end
    new_system(mdl); open_system(mdl);

    % State-space matrices
    A = [zeros(n), eye(n);
         -diag(omega.^2), zeros(n)];
    B = [zeros(n,1); Phi.'*[F0;0;0]];  % modal forcing gain for u=sin(Omega t)
    C = [ [1 0 0]*Phi, zeros(1,n) ];   % x1 depends only on q
    D = 0;

    % Add blocks
    add_block('simulink/Sources/Sine Wave',[mdl '/Sine'], 'Amplitude','1', 'Frequency',num2str(Omega));
    add_block('simulink/Continuous/State-Space',[mdl '/ModalSS']);
    set_param([mdl '/ModalSS'],'A','A','B','B','C','C','D','D');

    add_block('simulink/Sinks/Scope',[mdl '/Scope']);

    % Connect
    add_line(mdl,'Sine/1','ModalSS/1');
    add_line(mdl,'ModalSS/1','Scope/1');

    % Push matrices to base workspace so the block can see them
    assignin('base','A',A); assignin('base','B',B);
    assignin('base','C',C); assignin('base','D',D);

    set_param(mdl,'StopTime','8');
    save_system(mdl);
    disp(['Simulink model created: ' mdl '.slx']);
end

File: Tutorials/Codes/System_Dynamics_Course/Chapter13/Lesson3/Chapter13_Lesson3.py
"""
Chapter13_Lesson3.py
System Dynamics (Control Engineering) — Chapter 13, Lesson 3
Modal Coordinates and Decoupling of MDOF Systems (Undamped Case)

This script:
1) Builds a 3-DOF mass-spring chain (M, K)
2) Solves the generalized eigenproblem K phi = lambda M phi
3) Mass-normalizes modes so that Phi^T M Phi = I and Phi^T K Phi = Omega^2
4) Simulates forced response using modal coordinates (decoupled ODEs)
5) Validates against direct physical-coordinate simulation

Dependencies:
  numpy, scipy, matplotlib
"""
import numpy as np
from numpy.linalg import norm
from scipy.linalg import eigh
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt


def mass_normalize(Phi: np.ndarray, M: np.ndarray) -> np.ndarray:
    """Scale columns of Phi so that Phi^T M Phi = I."""
    Phi_n = Phi.copy()
    for i in range(Phi.shape[1]):
        mi = Phi[:, i].T @ M @ Phi[:, i]
        Phi_n[:, i] = Phi[:, i] / np.sqrt(mi)
    return Phi_n


def build_3dof_chain(m=(1.0, 1.2, 0.9), k=(1200.0, 900.0, 700.0, 1100.0)):
    """
    3 masses connected by 4 springs:
      wall --k1-- m1 --k2-- m2 --k3-- m3 --k4-- wall
    """
    m1, m2, m3 = m
    k1, k2, k3, k4 = k
    M = np.diag([m1, m2, m3])
    K = np.array([
        [k1 + k2,   -k2,       0.0],
        [-k2,     k2 + k3,   -k3],
        [0.0,       -k3,     k3 + k4]
    ], dtype=float)
    return M, K


def modal_matrices(M: np.ndarray, K: np.ndarray):
    """
    Solve K phi = lambda M phi (symmetric generalized EVP).
    Returns (omega, Phi_mass_norm, Omega2_diag)
    """
    # eigh returns eigenvalues in ascending order for symmetric generalized EVP
    lam, Phi = eigh(K, M)
    lam = np.maximum(lam, 0.0)
    omega = np.sqrt(lam)
    Phi = mass_normalize(Phi, M)
    Omega2 = Phi.T @ K @ Phi
    return omega, Phi, Omega2


def simulate_modal(omega, Phi, force_fun, t_span=(0.0, 5.0), x0=None, v0=None, n_eval=2000):
    """
    Modal ODE: qdd + Omega^2 q = p(t), where p(t) = Phi^T f(t) (mass-normalized).
    """
    n = Phi.shape[0]
    if x0 is None:
        x0 = np.zeros(n)
    if v0 is None:
        v0 = np.zeros(n)

    # modal initial conditions: q0 = Phi^T M x0, qd0 = Phi^T M v0
    # since mass-normalized: Phi^T M Phi = I, and x = Phi q
    # thus q0 = Phi^T M x0, etc.
    # compute M from orthogonality identity: M = (Phi^{-T}) (Phi^{-1}) is not stable;
    # instead pass M in directly if needed. Here we reconstruct q0 by least squares:
    # x0 = Phi q0 -> q0 = argmin ||Phi q - x0||; same for v0.
    q0, *_ = np.linalg.lstsq(Phi, x0, rcond=None)
    qd0, *_ = np.linalg.lstsq(Phi, v0, rcond=None)

    def ode(t, z):
        q = z[:n]
        qd = z[n:]
        f = force_fun(t)  # physical force
        p = Phi.T @ f      # modal force
        qdd = - (omega**2) * q + p
        return np.hstack([qd, qdd])

    t_eval = np.linspace(t_span[0], t_span[1], n_eval)
    z0 = np.hstack([q0, qd0])
    sol = solve_ivp(ode, t_span, z0, t_eval=t_eval, rtol=1e-8, atol=1e-10)
    q = sol.y[:n, :]
    x = Phi @ q
    return sol.t, x, q


def simulate_physical(M, K, force_fun, t_span=(0.0, 5.0), x0=None, v0=None, n_eval=2000):
    """
    Physical ODE: M xdd + K x = f(t)  (undamped).
    Convert to first-order and integrate.
    """
    n = M.shape[0]
    if x0 is None:
        x0 = np.zeros(n)
    if v0 is None:
        v0 = np.zeros(n)
    Minv = np.linalg.inv(M)

    def ode(t, z):
        x = z[:n]
        xd = z[n:]
        f = force_fun(t)
        xdd = Minv @ (f - K @ x)
        return np.hstack([xd, xdd])

    t_eval = np.linspace(t_span[0], t_span[1], n_eval)
    z0 = np.hstack([x0, v0])
    sol = solve_ivp(ode, t_span, z0, t_eval=t_eval, rtol=1e-8, atol=1e-10)
    x = sol.y[:n, :]
    return sol.t, x


def main():
    M, K = build_3dof_chain()

    omega, Phi, Omega2 = modal_matrices(M, K)
    print("Natural frequencies (rad/s):", omega)
    print("Check Phi^T M Phi ~ I:\n", Phi.T @ M @ Phi)
    print("Check Phi^T K Phi ~ diag(omega^2):\n", Omega2)

    F0 = 10.0
    Omega = omega[0] * 0.9  # excite near first mode

    def f(t):
        return np.array([F0 * np.sin(Omega * t), 0.0, 0.0])

    t_span = (0.0, 8.0)
    x0 = np.zeros(3)
    v0 = np.zeros(3)

    t, x_modal, q = simulate_modal(omega, Phi, f, t_span=t_span, x0=x0, v0=v0)
    t2, x_phys = simulate_physical(M, K, f, t_span=t_span, x0=x0, v0=v0)

    # Plot
    plt.figure()
    plt.plot(t, x_modal[0, :], label="x1 (modal)")
    plt.plot(t2, x_phys[0, :], "--", label="x1 (physical)")
    plt.xlabel("t [s]")
    plt.ylabel("Displacement [m]")
    plt.legend()
    plt.title("Validation: modal vs physical simulation (DOF 1)")

    plt.figure()
    for i in range(3):
        plt.plot(t, q[i, :], label=f"q{i+1}")
    plt.xlabel("t [s]")
    plt.ylabel("Modal coordinate")
    plt.legend()
    plt.title("Modal coordinates q(t)")

    plt.show()


if __name__ == "__main__":
    main()
